README.md
collectd-openstack
==================

## Overview

A set of collectd plugins monitoring and publishing metrics for OpenStack components.

## Screenshots

![image](https://raw.github.com/catalyst/collectd-openstack/master/public/openstack-usage.png)

## Plugins and Metrics

There are several plugins, mapping to each of the openstack components.

Find below a list of the available plugins and the metrics they publish.

* cinder_plugin
  * openstack-cinder.&lt;tenant>.volumes.count (number of tenant volumes)
  * openstack-cinder.&lt;tenant>.volumes.bytes (total bytes of tenant volumes)
  * openstack-cinder.&lt;tenant>.volume-snapshots.count (number of tenant snapshots)
  * openstack-cinder.&lt;tenant>.volume-snapshots.bytes (total bytes of tenant snapshots)
  * openstack-cinder.&lt;tenant>.limites-{maxTotalVolumeGigabytes,maxTotalVolumes}
* glance_plugin
  * openstack-glance.&lt;tenant>.images.count (number of tenant images)
  * openstack-glance.&lt;tenant>.images.bytes (total bytes of tenant images)
* keystone_plugin
  * openstack-keystone.&lt;tenant>.users.count (number of tenant users)
  * openstack-keystone.totals.tenants.count (total number of tenants)
  * openstack-keystone.totals.users.count (total number of users)
  * openstack-keystone.totals.services.count (total number of services)
  * openstack-keystone.totals.endpoints.count (total number of endpoints)
* neutron_plugin
  * openstack-neutron.&lt;tenant>.networks.count (number of tenant networks)
  * openstack-neutron.&lt;tenant>.subnets.count (number of tenant subnets)
  * openstack-neutron.&lt;tenant>.routers.count (number of tenant routers)
  * openstack-neutron.&lt;tenant>.ports.count (number of tenant ports)
  * openstack-neutron.&lt;tenant>.floatingips.count (number of tenant floating ips)
  * openstack-neutron.&lt;tenant>.quotas-{floatingip,ikepolicy,ipsec_site_connection,ipsecpolicy,network,
                                          port,router,security_group,security_group_rule,subnet}
* nova_plugin
  * openstack-nova.&lt;tenant>.limits.{maxImageMeta,maxPersonality,maxPersonalitySize,maxSecurityGroupRules,
                                       maxSecurityGroups,maxServerMeta,maxTotalCores,maxTotalFloatingIps,
                                       maxTotalInstances,maxTotalKeypairs,maxTotalRAMSize}
    (limits on each metric, per tenant)
  * openstack-nova.&lt;tenant>.quotas.{cores,fixed_ips,floating_ips,instances,key_pairs,,ram,security_groups}
    (quotas on each metric, per tenant)
  * openstack-nova.cluster.gauge.config-AllocationRatioCores (overcommit ratio for vcpus)
  * openstack-nova.cluster.gauge.config-AllocationRatioRamMB (overcommit ratio for ram - in MB)
  * openstack-nova.cluster.gauge.config-ReservedCores (reserved vcpus on the whole cluster)
  * openstack-nova.cluster.gauge.config-ReservedRamMB (reserved ram on the whole cluster - in MB)
  * openstack-nova.cluster.gauge.config-ReservedNodeCores (reserved cores per node)
  * openstack-nova.cluster.gauge.config-ReservedNodeRamMB (reserved ram per node - in MB)
  * openstack-nova.&lt;hypervisor-hostname>.{current_workload,free_disk_gb,free_ram_mb,hypervisor_version,
                                             memory_mb,memory_mb_overcommit,memory_mb_overcommit_withreserve,
                                             memory_mb_used,running_vms,vcpus,vcpus_overcommit,
                                             vcpus_overcommit_withreserve,vcpus_used}

## Requirements

It assumes an existing installation of [collectd](http://collectd.org/documentation.shtml).

Check its documentation for details.

## Setup and Configuration

The example configuration(s) below assume the plugins to be located under `/usr/lib/collectd/plugins`.

If you're under ubuntu, consider installing from [this ppa](https://launchpad.net/~catalystit/+archive/collectd).

Each plugin should have its own config file, under `/etc/collectd/conf.d/<pluginname>.conf`, which
should follow some similar to:
```
# cat /etc/collectd/conf.d/keystone.conf

<LoadPlugin "python">
    Globals true
</LoadPlugin>

<Plugin "python">
    ModulePath "/usr/lib/collectd/plugins"

    Import "keystone_plugin"

    <Module "keystone_plugin">
        Username "admin"
        Password "123456"
        TenantName "openstack"
        AuthURL "https://api.example.com:5000/v2.0"
        Verbose "False"
    </Module>
</Plugin>
```

### Puppet

If you use puppet for configuration, then try this excelent [collectd](https://github.com/pdxcat/puppet-module-collectd) module.

It has plenty of docs on how to use it, but for our specific plugins:
```
  collectd::plugin::python { 'keystone':
    modulepath => '/usr/lib/collectd/plugins',
    module     => 'keystone_plugin',
    config     => {
      'Username'   => 'admin',
      'Password'   => '123456',
      'TenantName' => 'openstack',
      'AuthURL'    => 'https://api1.example.com:5000/v2.0',
      'Verbose'    => 'true',
    },
  }
```

## Limitations

The debian packaging files are provided, but don't expect the deb in the official repos.

## Development

All contributions more than welcome, just send pull requests.

## License

GPLv2 (check LICENSE).

## Contributors

Ricardo Rocha <ricardo@catalyst.net.nz>

## Support

Please log tickets and issues at the [github home](https://github.com/catalyst/collectd-openstack/issues).

## Additional Notes

### Ubuntu Packaging

[These instructions](http://packaging.ubuntu.com/html/packaging-new-software.html) should give full details.

In summary, do this once to prepare your environment:
```
pbuilder-dist precise create
```

and for every release (from master):
```
mkdir /tmp/build-collectd-os
cd /tmp/build-collectd-os
wget https://github.com/catalyst/collectd-openstack/archive/master.zip
unzip master.zip
tar zcvf collectd-openstack-0.7.0.tar.gz collectd-openstack-master/
bzr dh-make collectd-openstack 0.7.0 collectd-openstack-0.7.0.tar.gz
cd collectd-openstack
bzr builddeb -S
cd ../build-area
pbuilder-dist precise build collectd-openstack_0.7.0-1ubuntu.dsc
dput ppa:catalystit/collectd ../collectd-openstack_0.7.0-1ubuntu_source.changes
```


debian/changelog
collectd-openstack (0.8.0-1ubuntu) precise; urgency=medium

  * Add support for multiple regions in keystone catalog

 -- Xav Paice <xav@catalyst.net.nz>  Tue, 09 Dec 2014 11:12:08 +1300

collectd-openstack (0.7.0-1ubuntu) precise; urgency=medium

  * Add NoTenants option to disable per-tenant metrics

 -- Xav Paice <xav@catalyst.net.nz>  Mon, 08 Dec 2014 14:51:07 +1300

collectd-openstack (0.6.0-1ubuntu) precise; urgency=low

  * Bugs fixed:
    - #13: nova maxTotalRAMSize value is bogus
    - #11: interval not taken into account
    - #10: enable decent verbose logging
    - #9: quota values incorrect
    - #8: floating ip data now showing correctly
    - #7: volumes data missing

 -- Ricardo Rocha <ricardo@catalyst.net.nz>  Fri, 25 Jul 2014 16:09:45 +1200

collectd-openstack (0.5.0-1ubuntu) precise; urgency=low

  * Bugs fixed:
    - #1: allow overcommit ratios (memory and cpu)
    - #6: allow reserved values (for memory and cpus)
  * Added interval configuration param
  * Updated metrics with overcommit/reserved values (check README)

 -- Ricardo Rocha <ricardo@catalyst.net.nz>  Tue, 22 Jul 2014 10:28:45 +1200

collectd-openstack (0.4.3-1ubuntu) precise; urgency=low

  * Fixed #4: protect against missing tenant_id in cinder/neutron.

 -- Ricardo Rocha <ricardo@catalyst.net.nz>  Fri, 18 Jul 2014 10:28:45 +1200

collectd-openstack (0.4.2-1ubuntu) precise; urgency=low

  * Fixed #3: neutron plugin crashes on missing quotas

 -- Ricardo Rocha <ricardo@catalyst.net.nz>  Tue, 15 Jul 2014 17:28:45 +1200

collectd-openstack (0.4.1-1ubuntu) precise; urgency=low

  * Added missing __init__.py file in plugins directory.

 -- Ricardo Rocha <ricardo@catalyst.net.nz>  Tue, 08 Jul 2014 15:54:45 +1200

collectd-openstack (0.4-1ubuntu) precise; urgency=low

  * Moved plugins to /usr/lib/collectd/plugins/openstack.

 -- Ricardo Rocha <ricardo@catalyst.net.nz>  Mon, 23 Jun 2014 15:54:45 +1200

collectd-openstack (0.3-1ubuntu1) precise; urgency=low

  * Update for new upstream release.

 -- Ricardo Rocha <ricardo@catalyst.net.nz>  Tue, 06 May 2014 15:13:20 +1200

collectd-openstack (0.2-1ubuntu1) precise; urgency=low

  * Added dependency on python-cinderclient.
  * Update for new upstream release.

 -- Ricardo Rocha <ricardo@catalyst.net.nz>  Tue, 06 May 2014 12:31:06 +0200

collectd-openstack (0.1-2ubuntu1) precise; urgency=low

  * Initial release.

 -- Ricardo Rocha <ricardo@catalyst.net.nz>  Fri, 02 May 2014 18:25:06 +0200


plugins/base.py
#!/usr/bin/env python
#
# vim: tabstop=4 shiftwidth=4

# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; only version 2 of the License is applicable.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
#
# Authors:
#   Ricardo Rocha <ricardo@catalyst.net.nz>
#
# About this plugin:
#   Helper object for all plugins.
#
# collectd:
#   http://collectd.org
# collectd-python:
#   http://collectd.org/documentation/manpages/collectd-python.5.shtml
#
from keystoneclient.v2_0 import Client as KeystoneClient

import collectd
import datetime
import traceback

class Base(object):

    def __init__(self):
        self.username = 'admin'
        self.password = '123456'
        self.tenant = 'openstack'
        self.auth_url = 'http://api.example.com:5000/v2.0'
        self.verbose = False
        self.debug = False
        self.prefix = ''
        self.interval = 60.0
        self.notenants = False

    def get_keystone(self):
        """Returns a Keystone.Client instance."""
        return KeystoneClient(username=self.username, password=self.password,
                tenant_name=self.tenant, auth_url=self.auth_url)

    def config_callback(self, conf):
        """Takes a collectd conf object and fills in the local config."""
        for node in conf.children:
            if node.key == "Username":
                self.username = node.values[0]
            elif node.key == "Password":
                self.password = node.values[0]
            elif node.key == "TenantName":
                self.tenant = node.values[0]
            elif node.key == "AuthURL":
                self.auth_url = node.values[0]
            elif node.key == "Verbose":
                if node.values[0] in ['True', 'true']:
                    self.verbose = True
            elif node.key == "Debug":
                if node.values[0] in ['True', 'true']:
                    self.debug = True
            elif node.key == "AllocationRatioCores":
                self.AllocationRatioCores = float(node.values[0])
            elif node.key == "AllocationRatioRam":
                self.AllocationRatioRam = float(node.values[0])
            elif node.key == "ReservedNodeCores":
                self.ReservedNodeCores = float(node.values[0])
            elif node.key == "ReservedNodeRamMB":
                self.ReservedNodeRamMB = float(node.values[0])
            elif node.key == "ReservedCores":
                self.ReservedCores = float(node.values[0])
            elif node.key == "ReservedRamMB":
                self.ReservedRamMB = float(node.values[0])
            elif node.key == "Prefix":
                self.prefix = node.values[0]
            elif node.key == 'Interval':
                self.interval = float(node.values[0])
            elif node.key == 'NoTenants':
                self.notenants = True
            else:
                collectd.warning("%s: unknown config key: %s" % (self.prefix, node.key))

    def dispatch(self, stats):
        """
        Dispatches the given stats.

        stats should be something like:

        {'plugin': {'plugin_instance': {'type': {'type_instance': <value>, ...}}}}
        """
        if not stats:
            collectd.error("%s: failed to retrieve stats" % self.prefix)
            return

        self.logdebug("dispatching %d new stats :: %s" % (len(stats), stats))
        try:
            for plugin in stats.keys():
                for plugin_instance in stats[plugin].keys():
                    for type in stats[plugin][plugin_instance].keys():
                        type_value = stats[plugin][plugin_instance][type]
                        if not isinstance(type_value, dict):
                            self.dispatch_value(plugin, plugin_instance, type, None, type_value)
                        else:
                          for type_instance in stats[plugin][plugin_instance][type].keys():
                                self.dispatch_value(plugin, plugin_instance,
                                        type, type_instance,
                                        stats[plugin][plugin_instance][type][type_instance])
        except Exception as exc:
            collectd.error("%s: failed to dispatch values :: %s :: %s"
                    % (self.prefix, exc, traceback.format_exc()))

    def dispatch_value(self, plugin, plugin_instance, type, type_instance, value):
        """Looks for the given stat in stats, and dispatches it"""
        self.logdebug("dispatching value %s.%s.%s.%s=%s"
                % (plugin, plugin_instance, type, type_instance, value))

        val = collectd.Values(type='gauge')
        val.plugin=plugin
        val.plugin_instance=plugin_instance
        if type_instance is not None:
            val.type_instance="%s-%s" % (type, type_instance)
        else:
            val.type_instance=type
        val.values=[value]
        val.interval = self.interval
        val.dispatch()
        self.logdebug("sent metric %s.%s.%s.%s.%s"
                % (plugin, plugin_instance, type, type_instance, value))

    def read_callback(self):
        try:
            start = datetime.datetime.now()
            stats = self.get_stats()
            self.logverbose("collectd new data from service :: took %d seconds"
                    % (datetime.datetime.now() - start).seconds)
        except Exception as exc:
            collectd.error("%s: failed to get stats :: %s :: %s"
                    % (self.prefix, exc, traceback.format_exc()))
        self.dispatch(stats)

    def get_stats(self):
        collectd.error('Not implemented, should be subclassed')

    def logverbose(self, msg):
        if self.verbose:
            collectd.info("%s: %s" % (self.prefix, msg))

    def logdebug(self, msg):
        if self.debug:
            collectd.info("%s: %s" % (self.prefix, msg))



plugins/cinder_plugin.py
#!/usr/bin/env python
#
# vim: tabstop=4 shiftwidth=4

# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; only version 2 of the License is applicable.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
#
# Authors:
#   Ricardo Rocha <ricardo@catalyst.net.nz>
#
# About this plugin:
#   This plugin collects OpenStack cinder information, including stats on
#   volumes and snapshots usage per tenant.
#
# TODO: Start using cinder v2 api
#
# collectd:
#   http://collectd.org
# OpenStack Cinder:
#   http://docs.openstack.org/developer/cinder
# collectd-python:
#   http://collectd.org/documentation/manpages/collectd-python.5.shtml
#
#!/usr/bin/env python
#
from cinderclient.client import Client as CinderClient

import collectd
import traceback

import base

class CinderPlugin(base.Base):

    def __init__(self):
        base.Base.__init__(self)
        self.prefix = 'openstack-cinder'

    def get_stats(self):
        """Retrieves stats from cinder"""
        keystone = self.get_keystone()

        tenant_list = keystone.tenants.list()

        tenants = {}
        data = { self.prefix: {} }

        # Initialize tenant struct
        for tenant in tenant_list:
            tenants[tenant.id] = tenant.name
            data[self.prefix]["tenant-%s" % tenant.name] = {
                'limits': {}, 'quotas': {},
                'volumes': { 'count': 0, 'bytes': 0 },
                'volume-snapshots': { 'count': 0, 'bytes': 0 }
            }

        if getattr(self, 'region') is None:
            client = CinderClient('1', self.username, self.password,
                                  self.tenant, self.auth_url)
        else:
            client = CinderClient('1', self.username, self.password,
                                  self.tenant, self.auth_url,
                                  region_name=self.region)
        # Collect limits for each tenant (quotas data is bogus in havana)
        for tenant in tenant_list:
            data_tenant = data[self.prefix]["tenant-%s" % tenant.name]
            # limits call in havana does not expose a tenant_id param :(
            if getattr(self, 'region') is None:
                client2 = CinderClient('1', self.username, self.password,
                                       tenant.name, self.auth_url)
            else:
                client = CinderClient('1', self.username, self.password,
                                      self.tenant, self.auth_url,
                                      region_name=self.region)
            try:
                limits = client2.limits.get().absolute
            except Exception:
                continue
            for limit in limits:
                if 'Giga' in limit.name:
                    limit.value = limit.value * 1024 * 1024 * 1024
                data_tenant['limits'][limit.name] = limit.value

        # Ideally we would use limits, but the usage values are only available
        # in icehouse (havana needs a full volumes list)
        # https://github.com/openstack/cinder/commit/608a8dd55300555fb1d17cdf53d08d331d7f5064
        volumes = client.volumes.list(search_opts={'all_tenants': 1})
        for volume in volumes:
            try:
                data_tenant = data[self.prefix]["tenant-%s" % tenants[getattr(volume, 'os-vol-tenant-attr:tenant_id')]]
            except KeyError:
                continue
            data_tenant['volumes']['count'] += 1
            data_tenant['volumes']['bytes'] += (volume.size * 1024 * 1024 * 1024)

        snapshots = client.volume_snapshots.list(search_opts={'all_tenants': 1})
        # TODO: seems the project_id extattr is not always available, need to check why
        # https://github.com/rochaporto/collectd-openstack/issues/12
        for snapshot in snapshots:
            try:
                tenant_id = getattr(snapshot, 'os-extended-snapshot-attributes:project_id')
            except AttributeError:
                continue
            try:
                data_tenant = data[self.prefix]["tenant-%s" % tenants[tenant_id]]
            except KeyError:
                continue
            data_tenant['volume-snapshots']['count'] += 1
            data_tenant['volume-snapshots']['bytes'] += (snapshot.size * 1024 * 1024 * 1024)

        return data

try:
    plugin = CinderPlugin()
except Exception as exc:
    collectd.error("openstack-cinder: failed to initialize cinder plugin :: %s :: %s"
            % (exc, traceback.format_exc()))

def configure_callback(conf):
    """Received configuration information"""
    plugin.config_callback(conf)

def read_callback():
    """Callback triggerred by collectd on read"""
    plugin.read_callback()

collectd.register_config(configure_callback)
collectd.register_read(read_callback, plugin.interval)


plugins/nova_plugin.py
#!/usr/bin/env python
#
# vim: tabstop=4 shiftwidth=4

# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; only version 2 of the License is applicable.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
#
# Authors:
#   Ricardo Rocha <ricardo@catalyst.net.nz>
#
# About this plugin:
#   This plugin collects OpenStack nova information, including limits and
#   quotas per tenant.
#
# collectd:
#   http://collectd.org
# OpenStack Nova:
#   http://docs.openstack.org/developer/nova
# collectd-python:
#   http://collectd.org/documentation/manpages/collectd-python.5.shtml
#
from novaclient.client import Client as NovaClient
from novaclient import exceptions

import collectd
import traceback

import base

class NovaPlugin(base.Base):

    def __init__(self):
        base.Base.__init__(self)
        self.prefix = 'openstack-nova'

    def get_stats(self):
        """Retrieves stats from nova"""
        keystone = self.get_keystone()

        data = { self.prefix: { 'cluster': { 'config': {} }, } }
        if getattr(self, 'region') is None:
            client = NovaClient('2', self.username, self.password, self.tenant, self.auth_url)
        else:
            client = NovaClient('2', self.username, self.password, self.tenant, self.auth_url,
                                region_name=self.region)

        if getattr(self, 'notenants') == False:
            tenant_list = keystone.tenants.list()

            for tenant in tenant_list:
                # FIX: nasty but works for now (tenant.id not being taken below :()
                client.tenant_id = tenant.id
                data[self.prefix]["tenant-%s" % tenant.name] = { 'limits': {}, 'quotas': {} }
                data_tenant = data[self.prefix]["tenant-%s" % tenant.name]

                # Get absolute limits for tenant
                limits = client.limits.get(tenant_id=tenant.id).absolute
                for limit in limits:
                    if 'ram' in limit.name.lower():
                        limit.value = limit.value * 1024.0 * 1024.0
                    data_tenant['limits'][limit.name] = limit.value

                # Quotas for tenant
                quotas = client.quotas.get(tenant.id)
                for item in ('cores', 'fixed_ips', 'floating_ips', 'instances',
                    'key_pairs', 'ram', 'security_groups'):
                    if item == 'ram':
                        setattr(quotas, item, getattr(quotas, item) * 1024 * 1024)
                    data_tenant['quotas'][item] = getattr(quotas, item)

        # Cluster allocation / reserved values
        for item in ('AllocationRatioCores', 'AllocationRatioRam',
                'ReservedNodeCores', 'ReservedNodeRamMB',
                'ReservedCores', 'ReservedRamMB'):
            data[self.prefix]['cluster']['config'][item] = getattr(self, item)

        # Hypervisor information
        hypervisors = client.hypervisors.list()
        for hypervisor in hypervisors:
            name = "hypervisor-%s" % hypervisor.hypervisor_hostname
            data[self.prefix][name] = {}
            for item in ('current_workload', 'free_disk_gb', 'free_ram_mb',
                    'hypervisor_version', 'memory_mb', 'memory_mb_used',
                    'running_vms', 'vcpus', 'vcpus_used'):
                data[self.prefix][name][item] = getattr(hypervisor, item)
            data[self.prefix][name]['memory_mb_overcommit'] = \
                data[self.prefix][name]['memory_mb'] * data[self.prefix]['cluster']['config']['AllocationRatioRam']
            data[self.prefix][name]['memory_mb_overcommit_withreserve'] = \
                data[self.prefix][name]['memory_mb_overcommit'] - data[self.prefix]['cluster']['config']['ReservedNodeRamMB']
            data[self.prefix][name]['vcpus_overcommit'] = \
                data[self.prefix][name]['vcpus'] * data[self.prefix]['cluster']['config']['AllocationRatioCores']
            data[self.prefix][name]['vcpus_overcommit_withreserve'] = \
                data[self.prefix][name]['vcpus_overcommit'] - data[self.prefix]['cluster']['config']['ReservedNodeCores']

        return data

try:
    plugin = NovaPlugin()
except Exception as exc:
    collectd.error("openstack-nova: failed to initialize nova plugin :: %s :: %s"
            % (exc, traceback.format_exc()))

def configure_callback(conf):
    """Received configuration information"""
    plugin.config_callback(conf)

def read_callback():
    """Callback triggerred by collectd on read"""
    plugin.read_callback()

collectd.register_config(configure_callback)
collectd.register_read(read_callback, plugin.interval)
